src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -14,78 +19,203 @@ module Lorentz.Contracts.BaseDAO.Types
+
+  , sProposals :: BigMap ProposalId (Proposal proposalMetadata)
+  , sProposalsIdSortByDate :: [ProposalId] -- Needed for `flush`
+
+  -- Configuration (TODO: make it a separate record)
+  , sRejectedProposalReturnValue :: Natural
   }
   deriving stock (Generic, Show)
-  deriving anyclass (IsoValue, HasAnnotation)
+  deriving anyclass (HasAnnotation)
 
-instance TypeHasDoc Storage where
-  typeDocMdDescription =
-    "Contract's storage holding a big_map with all balances and the operators."
+deriving anyclass instance (WellTypedIsoValue proposalMetadata) => IsoValue (Storage proposalMetadata)
 
-instance HasFieldOfType Storage name field =>
-         StoreHasField Storage name field where
+instance HasFieldOfType (Storage pm) name field =>
+         StoreHasField (Storage pm) name field where
   storeFieldOps = storeFieldOpsADT
 
-type StorageC store =
-  StorageContains store
-    [ "sAdmin" := Address
+instance IsoValue pm =>
+    StoreHasSubmap (Storage pm) "sLedger" LedgerKey LedgerValue where
+  storeSubmapOps = storeSubmapOpsDeeper #sLedger
+
+instance IsoValue pm =>
+    StoreHasSubmap (Storage pm) "sOperators" (Address, (Address, FA2.TokenId)) () where
+  storeSubmapOps = storeSubmapOpsDeeper #sOperators
+
+instance IsoValue pm =>
+    StoreHasSubmap (Storage pm) "sProposals" ProposalId (Proposal pm) where
+  storeSubmapOps = storeSubmapOpsDeeper #sProposals
+
+type StorageC store pm =
+  ( StorageContains store
+    [ "sLedger" := LedgerKey ~> LedgerValue
+    , "sOperators" := (Address, (Address, FA2.TokenId)) ~> ()
+    , "sTokenAddress" := Address
+    , "sAdmin" := Address
     , "sPendingOwner" := Maybe Address
     , "sLedger" := Ledger
+
+    , "sProposalIdCounter" := Natural
+
+    , "sProposals" := ProposalId ~> Proposal pm
+    , "sProposalsIdSortByDate" := [ProposalId]
+
+    , "sRejectedProposalReturnValue" := Natural
     ]
 
+  --  [ "sLedger" := LedgerKey ~> LedgerValue
+  --  , "sOperators" := (Address, (Address, FA2.TokenId)) ~> ()
+  --  , "sTokenAddress" := Address
+
+  --  , "sAdmin" := Address
+  --  , "sProposalIdCounter" := Natural
+
+  --  , "sProposals" := ProposalId ~> Proposal pm
+  --  , "sProposalsIdSortByDate" := [ProposalId]
+
+  --  , "sRejectedProposalReturnValue" := Natural
+  --  ]
+  )
+
 -- | Parameter of the BaseDAO contract
-data Parameter
+data Parameter proposalMetadata
   = Call_FA2 FA2.Parameter
   | Transfer_ownership TransferOwnershipParam
   | Accept_ownership ()
+  | Propose (ProposeParams proposalMetadata)
+  -- | Proposal_Metadata ()
+  -- | SetVotingPeriod ()
+  -- | SetQuorumThreshold ()
+  -- | Vote ()
+  -- | Flush ()
   deriving stock (Generic, Show)
-  deriving anyclass (IsoValue)
 
-instance ParameterHasEntrypoints Parameter where
-  type ParameterEntrypointsDerivation Parameter = EpdDelegate
+instance (HasAnnotation pm, IsoValue pm) => ParameterHasEntrypoints (Parameter pm) where
+  type ParameterEntrypointsDerivation (Parameter pm) = EpdRecursive
```

Comment:
Why not keep `EpdDelegate`?

Reply:
Ater adding the type params `proposalMetadata`, EpdDelegate no longer works:
```
    • Could not deduce: (Michelson.Typed.Scope.FailOnOperationFound
                           (Michelson.Typed.Scope.ContainsOp (ToT pm)),
                         Michelson.Typed.Scope.FailOnNestedBigMapsFound
                           (Michelson.Typed.Scope.ContainsNestedBigMaps (ToT pm)))
        arising from the superclasses of an instance declaration
      from the context: (HasAnnotation pm, IsoValue pm)
        bound by the instance declaration
        at src/Lorentz/Contracts/BaseDAO/Types.hs:184:10-82
    • In the instance declaration for
        ‘ParameterHasEntrypoints (Parameter pm)’
    |       
184 | instance (HasAnnotation pm, IsoValue pm) => ParameterHasEntrypoints (Parameter pm) where
```

Reply:
I've updated this, and all other comments as well, Ivan.